File: Entities/Bucket.cs
using System;
using System.Collections.Generic;
using SecretShare.Entities.Interfaces;

namespace SecretShare.Entities
{

    /// <summary>
    /// A bucket is a en entity that can contain several secrets.
    /// The bucket holds the public and private key for encrypting and decrypting.
    /// </summary>
    /// <value></value>
    public record Bucket : IDateTracked
    {

        /// <summary>
        /// The id of this bucket
        /// </summary>
        public Guid BucketId { get; init; }

        /// <summary>
        /// The id used to get this bucket for submission
        /// </summary>
        public Guid SubmissionId { get; init; } = Guid.NewGuid();

        /// <summary>
        /// The public key for this bucket, used for encrypting
        /// </summary>
        public string PublicKey { get; init; }

        /// <summary>
        /// The private key used to decrypt the secrets submitted.
        /// This key should be encrypted before stored.
        /// </summary>
        public string PrivateKey { get; init; }

        /// <summary>
        /// All secrets belonging to this bucket.
        /// </summary>
        public ICollection<Secret> Secrets { get; init; }
        public Bucket()
        {
            this.Secrets = new HashSet<Secret>();
        }

        public DateTimeOffset CreatedAt { get; set; }

        public DateTimeOffset UpdatedAt { get; set; }
    }
}

File: frontend/src/lib/models/Bucket.ts
import { BucketDto } from '../dtos/Response/bucketDto'
import { SecretDto } from '../dtos/Response/secretDto'

export class Bucket implements BucketDto {
  private _bucketId: string = ''
  public get bucketId(): string {
    return this._bucketId
  }
  public set bucketId(value: string) {
    this._bucketId = value
  }
  private _publicKey: string = ''
  public get publicKey(): string {
    return this._publicKey
  }
  public set publicKey(value: string) {
    this._publicKey = value
  }
  private _privateKey: string = ''
  public get privateKey(): string {
    return this._privateKey
  }
  public set privateKey(value: string) {
    this._privateKey = value
  }
  private _secrets: SecretDto[] = []
  public get secrets(): SecretDto[] {
    return this._secrets
  }
  public set secrets(value: SecretDto[]) {
    this._secrets = value
  }
  private _createdAt: Date = new Date()
  public get createdAt(): Date {
    return this._createdAt
  }
  public set createdAt(value: Date) {
    this._createdAt = value
  }
  private _updatedAt: Date = new Date()
  public get updatedAt(): Date {
    return this._updatedAt
  }
  public set updatedAt(value: Date) {
    this._updatedAt = value
  }
}
